test/test_device_radix_sort.cu

```diff
@@ -626,73 +628,218 @@ template <typename KeyT>
         }
         else
         {
-            h_pairs[i].key = h_keys[i];
+            memcpy(h_reference_keys, h_keys, num_items * sizeof(KeyT));
         }
 
-        h_pairs[i].value = i;
+        // Summarize the pre-sorted array (element, 1st position, count).
+        struct Element
+        {
+            KeyT key;
+            size_t num;
+            size_t index;
+        };
+        std::vector<Element> summary;
+        KeyT cur_key = h_reference_keys[0];
+        summary.push_back(Element{cur_key, 1, 0});
+        for (size_t i = 1; i < num_items; ++i)
+        {
+            // TODO: handle -0.0, +0.0 and NaNs correctly.
+            KeyT key = h_reference_keys[i];
+            if (key == cur_key)
+            {
+                // Same key.
+                summary.back().num++;
+                continue;
+            }
+
+            // Different key.
+            cur_key = key;
+            summary.push_back(Element{cur_key, 1, i});
+        }
+        
+        // Generate a random permutation from the summary. Such a complicated
+        // approach is used to permute the array and compute ranks in a
+        // cache-friendly way and in a short time.
+        h_reference_ranks = new size_t[num_items];
+        size_t max_run = 32, run = 0, i = 0;
+        while (summary.size() > 0)
+        {
+            // Pick up a random element and a run.
+            size_t bits = 0;
+            RandomBits(bits);
+            size_t j = bits % summary.size();
+            Element& element = summary[j];
+            run = std::min(1 + bits % (max_run - 1), element.num);
+            for (size_t j = 0; j < run; ++j)
+            {
+                h_keys[i + j] = element.key;
+                h_reference_ranks[element.index + j] = i + j;
+            }
+            i += run;
+            element.index += run;
+            element.num -= run;
+            if (element.num == 0)
+            {
+                // Remove the empty entry.
+                std::swap(summary[j], summary.back());
+                summary.pop_back();
+            }
+        }
+        printf(" Done.\n");
     }
+    else
+    {
+        typedef Pair<KeyT, int> PairT;
 
-    printf("\nSorting reference solution on CPU (%d segments)...", num_segments); fflush(stdout);
+        PairT *h_pairs = new PairT[num_items];
 
-    for (int i = 0; i < num_segments; ++i)
-    {
-        if (IS_DESCENDING) std::reverse(h_pairs + h_segment_offsets[i], h_pairs + h_segment_offsets[i + 1]);
-        std::stable_sort(               h_pairs + h_segment_offsets[i], h_pairs + h_segment_offsets[i + 1]);
-        if (IS_DESCENDING) std::reverse(h_pairs + h_segment_offsets[i], h_pairs + h_segment_offsets[i + 1]);
-    }
+        int num_bits = end_bit - begin_bit;
+        for (OffsetT i = 0; i < num_items; ++i)
+        {
 
-    printf(" Done.\n"); fflush(stdout);
+            // Mask off unwanted portions
+            if (num_bits < static_cast<int>(sizeof(KeyT) * 8))
```

Comment:
Should the presorted branch also account for begin/end bits?

Reply:
nvm, I see later on that we don't test bit ranges with pre-sorted inputs. Let's add a comment in the presorted branch to clarify this.

Reply:
I've added the comment.